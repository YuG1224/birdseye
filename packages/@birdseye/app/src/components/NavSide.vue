<template>
  <ul class="nav-side">
    <li
      v-for="decl in declarations"
      :key="decl.name"
      class="item">
      <router-link
        class="anchor"
        :class="{ current: isCurrent(decl.name) }"
        :to="previewRoute(decl.name)">
        {{ decl.name }}
      </router-link>

      <TransitionDisclosure name="child">
        <div
          v-show="component === decl.name"
          class="child-wrapper">
          <ul class="child-list">
            <li
              v-for="p in decl.patterns"
              :key="p.name"
              class="child-item">
              <router-link
                class="child-anchor"
                :class="{ current: isCurrent(decl.name, p.name) }"
                :to="previewRoute(decl.name, p.name)">
                {{ p.name }}
              </router-link>
            </li>
          </ul>
        </div>
      </TransitionDisclosure>
    </li>
  </ul>
</template>

<script lang="ts">
import Vue from 'vue'
import { Location } from 'vue-router'
import { ComponentDeclaration } from '@birdseye/core'
import { TransitionDisclosure } from 'vue-transition-components'

export default Vue.extend({
  name: 'NavSide',

  components: {
    TransitionDisclosure
  },

  props: {
    component: {
      type: String,
      default: null
    },

    pattern: {
      type: String,
      default: null
    },

    declarations: {
      type: Array as () => ComponentDeclaration[],
      required: true
    }
  },

  methods: {
    previewRoute(component: string, pattern?: string): Location {
      const to: Location = {
        name: 'preview',
        params: { component }
      }

      if (pattern) {
        to.params!.pattern = pattern
      }

      return to
    },

    isCurrent(component: string, pattern: string | null = null): boolean {
      return this.component === component && this.pattern === pattern
    }
  }
})
</script>

<style scoped>
.item {
  margin-bottom: 12px;
}

.item:last-child {
  margin-bottom: 0;
}

.anchor {
  display: block;
  font-weight: bold;
}

/**
 * Children
 */
.child-list {
  padding-top: 6px;
}

.child-item {
  margin-bottom: 6px;
}

.child-item:last-child {
  margin-bottom: 0;
}

.child-anchor {
  display: block;
  padding-left: 20px;
}

/**
 * Hover / Current
 */
.anchor:hover,
.anchor.current,
.child-anchor:hover,
.child-anchor.current {
  color: var(--color-main);
}

/**
 * Transition
 */
.child-enter-active,
.child-leave-active {
  transition: height 400ms var(--ease-out-cubic);
}
</style>
