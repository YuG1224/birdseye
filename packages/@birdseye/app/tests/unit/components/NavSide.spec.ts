import Router from 'vue-router'
import { mount, createLocalVue } from '@vue/test-utils'
import NavSide from '@/components/NavSide.vue'

describe('NavSide', () => {
  const declarations = [
    {
      name: 'Foo',
      patterns: [
        {
          name: 'Pattern 1'
        },
        {
          name: 'Pattern 2'
        }
      ]
    },
    {
      name: 'Bar',
      patterns: [
        {
          name: 'Pattern 1'
        }
      ]
    }
  ]

  let localVue: any

  beforeEach(() => {
    const router = new Router({
      routes: [
        {
          name: 'preview',
          path: '/:component/:pattern?',
          component: {},
          props: true
        }
      ]
    })
    localVue = createLocalVue().extend({ router })
    localVue.use(Router)
  })

  it('renders declarations', () => {
    const wrapper = mount(NavSide, {
      localVue,
      propsData: {
        declarations
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('highlight current component', () => {
    const wrapper = mount(NavSide, {
      localVue,
      propsData: {
        declarations,
        component: 'Foo'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('highlight current pattern', () => {
    const wrapper = mount(NavSide, {
      localVue,
      propsData: {
        declarations,
        component: 'Foo',
        pattern: 'Pattern 1'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
