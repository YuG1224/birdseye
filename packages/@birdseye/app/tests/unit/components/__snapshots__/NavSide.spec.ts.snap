// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavSide highlight current component 1`] = `
<ul class="nav-side">
  <li class="item">
    <a href="#/Foo" class="anchor current">
      Foo
    </a>
    <div class="child-wrapper" name="child">
      <ul class="child-list">
        <li class="child-item">
          <a href="#/Foo/Pattern%201" class="child-anchor">
            Pattern 1
          </a>
        </li>
        <li class="child-item">
          <a href="#/Foo/Pattern%202" class="child-anchor">
            Pattern 2
          </a>
        </li>
      </ul>
    </div>
  </li>
  <li class="item">
    <a href="#/Bar" class="anchor">
      Bar
    </a>
    <div class="child-wrapper" style="display: none;" name="child">
      <ul class="child-list">
        <li class="child-item">
          <a href="#/Bar/Pattern%201" class="child-anchor">
            Pattern 1
          </a>
        </li>
      </ul>
    </div>
  </li>
</ul>
`;

exports[`NavSide highlight current pattern 1`] = `
<ul class="nav-side">
  <li class="item">
    <a href="#/Foo" class="anchor">
      Foo
    </a>
    <div class="child-wrapper" name="child">
      <ul class="child-list">
        <li class="child-item">
          <a href="#/Foo/Pattern%201" class="child-anchor current">
            Pattern 1
          </a>
        </li>
        <li class="child-item">
          <a href="#/Foo/Pattern%202" class="child-anchor">
            Pattern 2
          </a>
        </li>
      </ul>
    </div>
  </li>
  <li class="item">
    <a href="#/Bar" class="anchor">
      Bar
    </a>
    <div class="child-wrapper" style="display: none;" name="child">
      <ul class="child-list">
        <li class="child-item">
          <a href="#/Bar/Pattern%201" class="child-anchor">
            Pattern 1
          </a>
        </li>
      </ul>
    </div>
  </li>
</ul>
`;

exports[`NavSide renders declarations 1`] = `
<ul class="nav-side">
  <li class="item">
    <a href="#/Foo" class="anchor">
      Foo
    </a>
    <div class="child-wrapper" style="display: none;" name="child">
      <ul class="child-list">
        <li class="child-item">
          <a href="#/Foo/Pattern%201" class="child-anchor">
            Pattern 1
          </a>
        </li>
        <li class="child-item">
          <a href="#/Foo/Pattern%202" class="child-anchor">
            Pattern 2
          </a>
        </li>
      </ul>
    </div>
  </li>
  <li class="item">
    <a href="#/Bar" class="anchor">
      Bar
    </a>
    <div class="child-wrapper" style="display: none;" name="child">
      <ul class="child-list">
        <li class="child-item">
          <a href="#/Bar/Pattern%201" class="child-anchor">
            Pattern 1
          </a>
        </li>
      </ul>
    </div>
  </li>
</ul>
`;
